service-proxy.js

```diff
@@ -112,8 +110,53 @@ function ServiceDispatchHandler(options) {
     self.peersToReap = Object.create(null);
     self.knownPeers = Object.create(null);
 
-    self.reapPeersTimer = null;
-    self.reapPeersPeriod = options.reapPeersPeriod || DEFAULT_REAP_PEERS_PERIOD;
+    self.peerReaper = new IntervalScan(
+        options.reapPeersPeriod || DEFAULT_REAP_PEERS_PERIOD,
+        function reapSinglePeer(hostPort, serviceNames) {
+            self.reapSinglePeer(hostPort, serviceNames);
+        },
+        function getPeersToReap() {
+            var peersToReap = self.peersToReap;
+            self.peersToReap = self.knownPeers;
+            self.knownPeers = Object.create(null);
+            return peersToReap;
+        },
+        self.channel.timers
+    );
+    self.peerReaper.runBeginEvent.on(function onPeerReapBegin(run) {
+        self.logger.info('reaping dead peers', self.extendLogInfo({
+            numPeersToReap: run.keys.length
+        }));
+    });
+    self.peerReaper.start(); // TODO: should IntervalScan be self-starting?
+
+    // ---
+
+    self.servicePurger = new IntervalScan(
+        options.servicePurgePeriod || SERVICE_PURGE_PERIOD,
+        function maybePurgeEachService(serviceName, lastRefresh) {
+            var now = self.channel.timers.now();
+            if (now - lastRefresh > self.servicePurgePeriod) {
+                delete self.exitServices[serviceName];
+                var chan = self.channel.subChannels[serviceName];
+                if (chan) {
+                    chan.close();
+                    delete self.channel.subChannels[serviceName];
+                    self.rateLimiter.removeServiceCounter(serviceName);
+                    self.rateLimiter.removeKillSwitchCounter(serviceName);
+                }
+            }
+        },
+        // TODO: would be convenient to be able to pass an object or a function
+        // in cases like this
+        function getExitServices() {
+            return self.exitServices;
```

Comment:
Added a guard to IntervalScanRun#nextItem to skip undefined entries, I'd rather not introduce a copy into places like this where they didn't exist before; also I actively want to get a, maybe newer, lastRefresh by the time I get around to calling `maybePurgeEachService`